addons/delivery/models/stock_picking.py

```diff
@@ -195,7 +195,13 @@ def _add_delivery_cost_to_so(self):
             delivery_lines = sale_order.order_line.filtered(lambda l: l.is_delivery and l.currency_id.is_zero(l.price_unit) and l.product_id == self.carrier_id.product_id)
             carrier_price = self.carrier_price * (1.0 + (float(self.carrier_id.margin) / 100.0))
             if not delivery_lines:
-                sale_order._create_delivery_line(self.carrier_id, carrier_price)
+                new_delivery_line = sale_order._create_delivery_line(self.carrier_id, carrier_price)
+                if self.backorder_id:
```

Comment:
You could just remove the else no?
and use delivery_line = [sale_order._create_delivery_line(self.carrier_id, carrier_price)]